### PR TITLE
Add RFC 7009 token revocation endpoint

### PR DESCRIPTION
**Pull request by indeewari**
Introduce POST /oauth2/revoke and the write path for the single-token deny list. This is the write side only — enforcement (rejecting a revoked token on the introspection/refresh/exchange hot paths) lands separately.

- revocation package: RFC 7009 service, handler, and route registration behind the clientauth middleware, plus the deny-list writer (InsertRevokedToken). The writer is unexported and kept separate from the read/enforcement path so the AS hot path can check the deny list but never write to it.
- Signature-only validation (VerifyJWTSignature): expired tokens remain revocable; an invalid/unknown/jti-less token is a no-op success per §2.2.
- Ownership via the token's client_id claim; a mismatch returns invalid_grant.
- Idempotent write (ON CONFLICT (DEPLOYMENT_ID, JTI) DO NOTHING).
- Discovery advertises revocation_endpoint; unsupported_token_type error and TOKEN_REVOKED audit event added; wired in oauth/init.

Unit tests cover the service, handler HTTP mapping, route registration, and writer (92% package coverage).

Refs https://github.com/thunder-id/thunderid/discussions/3321

### Purpose
Implementing the token revocation endpoint.

---
#### 🔧 Summary of Breaking Changes
_Describe what is changing_

#### 💥 Impact
_What will break? Who is affected?_

#### 🔄 Migration Guide
_How should users update their code/configuration to adapt to the breaking changes? Include examples if helpful_

---

-->

### Approach
Introduce POST /oauth2/revoke and the write path for the single-token deny list. This is the write side only — enforcement (rejecting a revoked token on the introspection/refresh/exchange hot paths) lands separately.

### Related Issues
- https://github.com/thunder-id/thunderid/issues/3603

### Related PRs
- N/A

### Checklist
- [X] Followed the contribution guidelines.
- [X] Manual test round performed and verified.
- [ ] Documentation provided. (Add links if there are any)
    - [ ] Ran Vale and fixed all errors and warnings
- [ ] Tests provided. (Add links if there are any)
    - [X] Unit Tests
    - [ ] Integration Tests
- [ ] Breaking changes. (Fill if applicable)
    - [ ] Breaking changes section filled.
    - [ ] `breaking change` label added.

### Security checks
- [X] Followed secure coding standards in [WSO2 Secure Coding Guidelines](https://security.docs.wso2.com/en/latest/security-guidelines/secure-engineering-guidelines/secure-coding-guidlines/introduction/)
- [X] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.

## Summary by CodeRabbit

* **New Features**
  * Added OAuth 2.0 token revocation (RFC 7009) via `POST /oauth2/revoke` (with CORS `OPTIONS`), including request handling, deny-list storage, and `TOKEN_REVOKED` observability events.
  * OAuth2 discovery metadata now advertises a non-empty `revocation_endpoint`.
* **Bug Fixes**
  * Added support for the OAuth2 error code `unsupported_token_type`.
* **Tests**
  * Added unit and integration coverage for revocation routing, handler outcomes (including edge cases), discovery metadata, and revoked-token storage behavior.

### CODE REVIEW

File: backend/internal/oauth/oauth2/constants/constants.go
Comment (Malith-19) [inline backend/internal/oauth/oauth2/constants/constants.go]: Let's update the license headers. Check the other places as well.

File: backend/internal/oauth/oauth2/revocation/handler.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+		return
+	}
+	tokenTypeHint := r.FormValue(constants.RequestParamTokenTypeHint)
+
+	clientID := ""
+	if client := clientauth.GetOAuthClient(ctx); client != nil {
+		clientID = client.ClientID
+	}
+
+	outcome, err := h.service.RevokeToken(ctx, token, tokenTypeHint, clientID)
```
Comment (Malith-19) [inline backend/internal/oauth/oauth2/revocation/handler.go]: Shall we name this to revocationResult or other meaningful name here?

File: backend/internal/oauth/oauth2/revocation/handler.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		h.logger.Error(ctx, "Failed to revoke token", log.Error(err))
+		sysutils.WriteJSONError(ctx, w, constants.ErrorServerError,
+			"An unexpected error occurred while processing the request",
+			http.StatusInternalServerError, nil)
+		return
+	}
+
+	switch outcome {
+	case RevokeOutcomeNotOwned:
+		sysutils.WriteJSONError(ctx, w, constants.ErrorInvalidGrant,
```
Comment (Malith-19) [inline backend/internal/oauth/oauth2/revocation/handler.go:79]: Do we need to add a debug log here?

File: backend/internal/oauth/oauth2/revocation/model.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	// ID is the surrogate UUID v7 primary key. Generated by the store when empty.
+	ID string
+	// JTI is the jti claim of the revoked token and the deny-list lookup key.
+	JTI string
+	// RevocationReason records why the token was revoked.
+	RevocationReason RevocationReason
+	// RevokedAt is the time the revocation was recorded.
+	RevokedAt time.Time
+	// ExpiryTime is the revoked token's original expiry; the row is removable once this passes.
+	ExpiryTime time.Time
```
Comment (Malith-19) [inline backend/internal/oauth/oauth2/revocation/model.go:47]: What is the usage of the expiryTime here?

File: backend/internal/oauth/oauth2/revocation/service.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+		return RevokeOutcomeRevoked, nil
+	}
+
+	_, payload, decodeErr := jwt.DecodeJWT(token)
+	if decodeErr != nil {
+		logger.Debug(ctx, "Revocation request for an undecodable token; treating as a no-op success")
+		return RevokeOutcomeRevoked, nil
+	}
+
+	jti, _ := payload["jti"].(string)
```
Comment (Malith-19) [inline backend/internal/oauth/oauth2/revocation/service.go]: Shall we use a constant for the "jti"

File: backend/internal/oauth/oauth2/revocation/model.go (hunk 3 above)
Comment (indeewari) [reply backend/internal/oauth/oauth2/revocation/model.go:47]: Expiry time is kept to handle cleanup in revoked token data.

File: backend/internal/oauth/oauth2/revocation/handler.go (hunk 2 above)
Comment (indeewari) [reply backend/internal/oauth/oauth2/revocation/handler.go:79]: Already covered at Debug in service.go:115; adding one in the handler would be redundant.

File: backend/internal/oauth/oauth2/constants/constants.go
Comment (indeewari) [reply backend/internal/oauth/oauth2/constants/constants.go]: Fixed

File: backend/internal/oauth/oauth2/revocation/handler.go (hunk 1 above)
Comment (indeewari) [reply backend/internal/oauth/oauth2/revocation/handler.go]: Fixed

File: backend/internal/oauth/oauth2/revocation/service.go (hunk 4 above)
Comment (indeewari) [reply backend/internal/oauth/oauth2/revocation/service.go]: Fixed

File: backend/internal/oauth/oauth2/revocation/handler.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+func newRevocationHandler(service RevocationServiceInterface) *revocationHandler {
+	return &revocationHandler{
+		service: service,
+		logger:  log.GetLogger().With(log.String(log.LoggerKeyComponentName, "RevocationHandler")),
+	}
+}
+
+// HandleRevoke handles token revocation requests. Client authentication is enforced upstream by the
+// clientauth middleware (401 invalid_client on failure).
+func (h *revocationHandler) HandleRevoke(w http.ResponseWriter, r *http.Request) {
```
Comment (rajithacharith) [inline backend/internal/oauth/oauth2/revocation/handler.go:46]: Is this request form Encoded or json encoded? Does the spec defines something for that?

Comment (indeewari) [reply backend/internal/oauth/oauth2/revocation/handler.go:46]: POST body url encoded, here it decodes first. url-encoding is mandated in spec.

File: backend/internal/oauth/oauth2/revocation/service.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+// RevokeToken validates and revokes the presented token. The token_type_hint parameter is ignored
+// (blank identifier) for the reasons documented on RevocationServiceInterface.RevokeToken.
+//
+// Per RFC 7009: signature is verified but expiry is intentionally not checked (expired tokens remain
+// revocable). An invalid, unparseable, or unknown token is a successful no-op. A token issued to a
+// different client is rejected with invalid_grant.
+func (s *revocationService) RevokeToken(
+	ctx context.Context, token, _, authenticatedClientID string,
```
Comment (rajithacharith) [inline backend/internal/oauth/oauth2/revocation/service.go:86]: Why do we have a _ ?

**Review by rajithacharith — APPROVED** [review]

**Comment by indeewari** [comment]
Closing as duplicate of 

- https://github.com/thunder-id/thunderid/pull/3727
